server/athenian/api/align/queries/metrics.py

```diff
@@ -192,185 +335,91 @@ def _jirafy_teams(
-    )
+    def collect(self, batch_calc_results: Sequence[np.ndarray]) -> TeamMetricsResult:
+        """Collect the result converting it into a structured TeamMetricsResult.
 
+        `batch_calc_results` must be obtained by calling the batch calc method with
+        the requests in `get_requests()`.
+        """
+        res: dict[Interval, dict[str, dict[int, object]]] = {}
+        for (request, team_ids), calc_result in zip(self._requests, batch_calc_results):
+            for interval_i, interval in enumerate(request.time_intervals):
+                int_ = cast(Interval, interval)
+                res.setdefault(int_, {})
+                for metric_i, metric in enumerate(request.metrics):
+                    res[int_].setdefault(metric, {})
+                    for team_i, team_id in enumerate(team_ids):
+                        val = self.get_val(calc_result, team_i, interval_i, metric_i)
+                        res[int_][metric][team_id] = val
+        return res
 
-def _triage_metric_values(
-    pr_metrics: Sequence[str],
-    release_metrics: Sequence[str],
-    jira_metrics: Sequence[str],
-    teams: Sequence[int],
-    time_intervals: Sequence[Interval],
-    metric_values: Tuple[Union[np.ndarray, Tuple[np.ndarray, ...]], ...],
-) -> Dict[Interval, Dict[str, Dict[int, object]]]:
-
-    pr_metric_values = release_metric_values = jira_metric_values = []
-    if pr_metrics:
-        pr_metric_values, *metric_values = metric_values
-    if release_metrics:
-        (release_metric_values, *_), *metric_values = metric_values
-    if jira_metrics:
-        (jira_metric_values, *_), *metric_values = metric_values
-
-    result: Dict[Interval, dict] = {}
-
-    for interval_idx, interval in enumerate(time_intervals):
-        result[interval] = {}
-        for metric_idx, metric in enumerate(pr_metrics):
-            metric_teams = result[interval][metric] = {}
-            for team, team_metric_values in zip(teams, pr_metric_values[0][0]):
-                metric_teams[team] = team_metric_values[interval_idx][0][metric_idx].value
-
-        for i, metric in enumerate(release_metrics):
-            metric_teams = result[interval][metric] = {}
-            for team, team_metric_values in zip(teams, release_metric_values):
-                metric_teams[team] = team_metric_values[0][interval_idx][0][i].value
-
-        for i, metric in enumerate(jira_metrics):
-            metric_teams = result[interval][metric] = {}
-            for team, team_metric_values in zip(teams, jira_metric_values):
-                metric_teams[team] = team_metric_values[0][interval_idx][0][i].value
+    def add_request(self, request: _MetricsLineRequestGenT, team_ids: Sequence[int]) -> None:
+        self._requests.append((request, team_ids))
 
-    return result
+    def get_requests(self) -> Sequence[_MetricsLineRequestGenT]:
+        return tuple(req[0] for req in self._requests)
```

Comment:
We should make as few calls to `calc()` as possible.

Every request is:
- One metric.
- Two time intervals (previous and current). Metrics calculated on wrong time intervals are a pure waste of resources.
- People. The actual participants are in different formats but always derive from the same set. Metrics calculated on wrong people are useless. `people = set[str]` 

Build the two-level dictionary (two time intervals) -> people -> list[metric]

Deduplicate list[metric].

Turn each list to a tuple, then inverse the bottom level: (two time intervals) -> tuple[metric] -> list[people]

Deduplicate list[people]. We've reached a reasonably good grouping. Each tree path is a separate call to `calc()`.

Reply:
done